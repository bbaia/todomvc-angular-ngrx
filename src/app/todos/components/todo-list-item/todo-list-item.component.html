<li
  *ngIf="todo as todo"
  [class.completed]="todo.completed"
  [class.editing]="editing"
>
  <div class="view">
    <input
      class="toggle"
      type="checkbox"
      [checked]="todo.completed"
      (click)="toggle.emit(todo.id)"
    />
    <label (dblclick)="activeEditMode()">
      {{ todo.text }}
    </label>
    <button class="destroy" (click)="delete.emit(todo.id)"></button>
  </div>
  <input
    class="edit"
    type="text"
    #textInput
    [hidden]="editing"
    [value]="todo.text"
    (keyup.enter)="updateText(todo.id, textInput.value)"
    (blur)="updateText(todo.id, textInput.value)"
  />
</li>
