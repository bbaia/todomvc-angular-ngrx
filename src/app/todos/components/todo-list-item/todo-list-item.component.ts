import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Todo } from './../../models';

@Component({
  selector: 'app-todo-list-item',
  templateUrl: './todo-list-item.component.html',
})
export class TodoListItemComponent {
  @Input() todo: Todo | null = null;
  @Output() toggle = new EventEmitter<number>();
  @Output() update = new EventEmitter<{ id: number; text: string }>();
  @Output() delete = new EventEmitter<number>();

  // 'local state is fine'
  editing = false;

  updateText(todoId: number, text: string): void {
    if (text && text.trim() !== this.todo?.text) {
      this.update.emit({ id: todoId, text: text.trim() });
    }
    this.editing = false;
  }

  activeEditMode(): void {
    this.editing = true;
  }
}
