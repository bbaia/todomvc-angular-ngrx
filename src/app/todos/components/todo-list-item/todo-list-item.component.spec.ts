import { ComponentFixture, TestBed } from '@angular/core/testing';
import { TodoListItemComponent } from './todo-list-item.component';

describe('TodoListItemComponent', () => {
  let component: TodoListItemComponent;
  let fixture: ComponentFixture<TodoListItemComponent>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [TodoListItemComponent],
    });
    fixture = TestBed.createComponent(TodoListItemComponent);
    component = fixture.componentInstance;
  });

  it('should create', () => {
    expect(component).toBeDefined();
  });

  it('should render', () => {
    component.todo = {
      id: 0.1,
      text: 'Learn TDD',
      creationDate: new Date(),
      completed: true,
    };
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should raise "toggle" event', done => {
    component.todo = {
      id: 0.1,
      text: 'Learn TDD',
      creationDate: new Date(),
      completed: false,
    };
    fixture.detectChanges();

    component.toggle.subscribe((id: number) => {
      // should trim text
      expect(id).toBe(0.1);
      done();
    });
    fixture.debugElement.nativeElement.querySelector('.toggle').click();
  });

  it('should raise "update" event', done => {
    component.todo = {
      id: 0.1,
      text: 'Learn TDD',
      creationDate: new Date(),
      completed: false,
    };
    fixture.detectChanges();

    component.update.subscribe(({ id, text }) => {
      // should trim text
      expect(id).toBe(0.1);
      expect(text).toEqual('Learn Jest');
      done();
    });
    fixture.debugElement.nativeElement
      .querySelector('.view > label')
      .dispatchEvent(
        new MouseEvent('dblclick', {
          bubbles: true,
          view: window,
        }),
      );
    const input: HTMLInputElement =
      fixture.debugElement.nativeElement.querySelector('.edit');
    input.value = 'Learn Jest ';
    input.dispatchEvent(new Event('input'));
    input.dispatchEvent(new Event('blur'));
  });

  it('should raise "delete" event', done => {
    component.todo = {
      id: 0.1,
      text: 'Learn TDD',
      creationDate: new Date(),
      completed: false,
    };
    fixture.detectChanges();

    component.delete.subscribe((id: number) => {
      // should trim text
      expect(id).toBe(0.1);
      done();
    });
    fixture.debugElement.nativeElement.querySelector('.destroy').click();
  });
});
