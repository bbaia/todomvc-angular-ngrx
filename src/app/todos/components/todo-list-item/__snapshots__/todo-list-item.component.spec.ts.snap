// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListItemComponent should render 1`] = `
<app-todo-list-item
  __ngContext__={[Function LRootView]}
  delete={[Function EventEmitter_]}
  editing="false"
  todo={[Function Object]}
  toggle={[Function EventEmitter_]}
  update={[Function EventEmitter_]}
>
  <li
    class="completed"
  >
    <div
      class="view"
    >
      <input
        class="toggle"
        type="checkbox"
      />
      <label>
         Learn TDD 
      </label>
      <button
        class="destroy"
      />
    </div>
    <input
      class="edit"
      type="text"
    />
  </li>
</app-todo-list-item>
`;
